### PR TITLE
[SWY-65] Add set section action menu UI

### PR DESCRIPTION
**Pull request by justaddcl**
## Background
This PR is to add the action menu to allow the user to perform actions on/against a set section
SSWY-65__after

## What's changed

- **New Features**
  - Set sections now feature an interactive menu that allows users to manage sections with options like reordering, editing, and deletion.
  - Song management actions have been streamlined for a smoother experience with enhanced dropdown interactions.
  - New properties added to components to support action menus and section management.
  
- **Bug Fixes**
  - Improved handling of user data loading and error states in action menus.

### Test Setup
1. Navigate to any set detail page
2. Verify that each set section card now displays a three-dots menu icon

#### Action Menu Visibility
1. Click the three-dots icon on any section
2. Verify that the action menu appears with options for:
   - Change section type
   - Move section up
   - Move section down
   - Delete section

#### Section Position Constraints
1. For the first section:
   - Verify "Move section up" is disabled but visible
2. For the last section:
   - Verify "Move section down" is disabled but visible
3. For middle sections:
   - Verify both move options are enabled and visible

#### Menu Interaction
1. Verify menu closes when:
   - Clicking outside the menu
   - Clicking the three-dots icon again
   - Clicking any menu item

#### Visual Positioning
1. Verify menu remains properly positioned when:
   - Page is scrolled
   - Window is resized
2. Verify menu appears aligned with the three-dots icon
3. Verify menu items are properly aligned and spaced

Note: This PR only implements the UI elements. The actual functionality for these actions will be implemented in future PRs.

### CODE REVIEW

**Comment by justaddcl** [comment]
#### 📝 Changed routes:


| [**…nization\]/sets/\[setId\]**](https://sanbi-g2e01vrnm-justaddcls-projects.vercel.app/[organization]/sets/[setId]) |
|:-:|
|<a href="https://sanbi-g2e01vrnm-justaddcls-projects.vercel.app/[organization]/sets/[setId]">Screenshot of /[organization]/sets/[setId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/qHc29iW7E9Mmkead1t4dU222XaX0pxT6fLGtCoOkzdI1cZS_kSHXV4-Ppt0S4Fwkz40rFj5KSZPvJTmmcBIhPA.png">(view full size)</a> |







Commit 9575f2c855dfe5140419d82f6230fdb426b1a2f8 (https://sanbi-g2e01vrnm-justaddcls-projects.vercel.app).

**Comment by justaddcl** [comment]
@coderabbitai Can you generate appropriately thorough unit tests for the `ActionMenu` and `ActionMenuItem` components that are production-ready?

**Comment by justaddcl** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> 🧩 Analysis chain
> @justaddcl Here are the production-ready unit tests for both components:
> 
> ```ts
> import { render, screen, fireEvent } from "@testing-library/react";
> import { ActionMenu, ActionMenuItem } from "./ActionMenu";
> import { vi } from "vitest";
> 
> describe("ActionMenu", () => {
>   const defaultProps = {
>     isOpen: false,
>     setIsOpen: vi.fn(),
>   };
> 
>   it("renders trigger button", () => {
>     render(<ActionMenu {...defaultProps}>Test Content</ActionMenu>);
>     
>     expect(screen.getByRole("button")).toBeInTheDocument();
>   });
> 
>   it("calls setIsOpen when trigger is clicked", () => {
>     render(<ActionMenu {...defaultProps}>Test Content</ActionMenu>);
>     
>     fireEvent.click(screen.getByRole("button"));
>     expect(defaultProps.setIsOpen).toHaveBeenCalled();
>   });
> 
>   it("renders children when open", () => {
>     render(
>       <ActionMenu isOpen={true} setIsOpen={vi.fn()}>
>         <div data-testid="test-content">Test Content</div>
>       </ActionMenu>
>     );
>     
>     expect(screen.getByTestId("test-content")).toBeInTheDocument();
>   });
> });
> 
> describe("ActionMenuItem", () => {
>   const mockOnClick = vi.fn();
>   const defaultProps = {
>     label: "Test Action",
>     icon: "Pencil" as const,
>     onClick: mockOnClick,
>   };
> 
>   it("renders with label and icon", () => {
>     render(<ActionMenuItem {...defaultProps} />);
>     
>     expect(screen.getByText("Test Action")).toBeInTheDocument();
>     expect(screen.getByLabelText("Test Action action")).toBeInTheDocument();
>   });
> 
>   it("handles click events", () => {
>     render(<ActionMenuItem {...defaultProps} />);
>     
>     fireEvent.click(screen.getByLabelText("Test Action action"));
>     expect(mockOnClick).toHaveBeenCalled();
>   });
> 
>   it("applies destructive styling", () => {
>     render(<ActionMenuItem {...defaultProps} destructive />);
>     
>     const container = screen.getByLabelText("Test Action action");
>     expect(container).toHaveClass("text-red-500");
>   });
> 
>   it("disables interaction when disabled", () => {
>     render(<ActionMenuItem {...defaultProps} disabled />);
>     
>     const container = screen.getByLabelText("Test Action action");
>     expect(container).toHaveClass("text-slate-400");
>     expect(container.closest('[disabled]')).toBeInTheDocument();
>   });
> 
>   it("renders different icons correctly", () => {
>     const { rerender } = render(<ActionMenuItem {...defaultProps} icon="Trash" />);
>     expect(screen.getByLabelText("Test Action action")).toBeInTheDocument();
> 
>     rerender(<ActionMenuItem {...defaultProps} icon="ArrowUp" />);
>     expect(screen.getByLabelText("Test Action action")).toBeInTheDocument();
>   });
> });
> ```
> 
> These tests cover:
> 
> 1. Basic rendering
> 2. Interaction handling
> 3. State management
> 4. Visual states
> 5. Icon rendering
> 6. Accessibility attributes
> 
> The tests follow the existing project patterns and include assertions for all main component features. You'll need to ensure you have `@testing-library/react` and `vitest` in your dev dependencies.

@coderabbitai My tests use jest, not vitest. Can you please amend your suggestions accordingly?

**Comment by justaddcl** [comment]
@coderabbitai Given the changes in this PR, can you generate thorough instructions on how to test the changes, but only for the scope of this PR?

**Comment by justaddcl** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> @justaddcl Here are the testing instructions for the set section action menu changes:
> 
> ### Test Setup
> 1. Navigate to any set detail page
> 2. Verify that each set section card now displays a three-dots menu icon
> 
> ### Test Cases
> #### Action Menu Visibility
> 1. Click the three-dots icon on any section
> 2. Verify that the action menu appears with options for:
>    
>    * Change section type
>    * Move section up
>    * Move section down
>    * Delete section
> 
> #### Section Position Constraints
> 1. For the first section:
>    
>    * Verify "Move section up" is disabled
> 2. For the last section:
>    
>    * Verify "Move section down" is disabled
> 3. For middle sections:
>    
>    * Verify both move options are enabled
> 
> #### Menu Actions
> 1. Test "Move section up":
>    
>    * Click on a non-first section's menu
>    * Select "Move section up"
>    * Verify the section moves one position up
> 2. Test "Move section down":
>    
>    * Click on a non-last section's menu
>    * Select "Move section down"
>    * Verify the section moves one position down
> 3. Test "Change section type":
>    
>    * Click menu on any section
>    * Select "Change section type"
>    * Verify section type options appear
> 4. Test "Delete section":
>    
>    * Click menu on any section
>    * Select "Delete section"
>    * Verify the section is removed from the set
> 
> #### Edge Cases
> 1. Verify menu closes when:
>    
>    * Clicking outside the menu
>    * Selecting an action
>    * Clicking the three-dots icon again
> 2. Verify menu remains properly positioned when:
>    
>    * Page is scrolled
>    * Window is resized

@coderabbitai The actual actions/mutations that happen after a user clicks the action menu items haven't been implemented yet. This PR is just to add the UI. Can you amend your test steps to accurately reflect that?

**Comment by justaddcl** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> @justaddcl Here are the updated tests using Jest:
> 
> ```ts
> import { render, screen, fireEvent } from "@testing-library/react";
> import { ActionMenu, ActionMenuItem } from "./ActionMenu";
> 
> describe("ActionMenu", () => {
>   const defaultProps = {
>     isOpen: false,
>     setIsOpen: jest.fn(),
>   };
> 
>   it("renders trigger button", () => {
>     render(<ActionMenu {...defaultProps}>Test Content</ActionMenu>);
>     
>     expect(screen.getByRole("button")).toBeInTheDocument();
>   });
> 
>   it("calls setIsOpen when trigger is clicked", () => {
>     render(<ActionMenu {...defaultProps}>Test Content</ActionMenu>);
>     
>     fireEvent.click(screen.getByRole("button"));
>     expect(defaultProps.setIsOpen).toHaveBeenCalled();
>   });
> 
>   it("renders children when open", () => {
>     render(
>       <ActionMenu isOpen={true} setIsOpen={jest.fn()}>
>         <div data-testid="test-content">Test Content</div>
>       </ActionMenu>
>     );
>     
>     expect(screen.getByTestId("test-content")).toBeInTheDocument();
>   });
> });
> 
> describe("ActionMenuItem", () => {
>   const mockOnClick = jest.fn();
>   const defaultProps = {
>     label: "Test Action",
>     icon: "Pencil" as const,
>     onClick: mockOnClick,
>   };
> 
>   it("renders with label and icon", () => {
>     render(<ActionMenuItem {...defaultProps} />);
>     
>     expect(screen.getByText("Test Action")).toBeInTheDocument();
>     expect(screen.getByLabelText("Test Action action")).toBeInTheDocument();
>   });
> 
>   it("handles click events", () => {
>     render(<ActionMenuItem {...defaultProps} />);
>     
>     fireEvent.click(screen.getByLabelText("Test Action action"));
>     expect(mockOnClick).toHaveBeenCalled();
>   });
> 
>   it("applies destructive styling", () => {
>     render(<ActionMenuItem {...defaultProps} destructive />);
>     
>     const container = screen.getByLabelText("Test Action action");
>     expect(container).toHaveClass("text-red-500");
>   });
> 
>   it("disables interaction when disabled", () => {
>     render(<ActionMenuItem {...defaultProps} disabled />);
>     
>     const container = screen.getByLabelText("Test Action action");
>     expect(container).toHaveClass("text-slate-400");
>     expect(container.closest('[disabled]')).toBeInTheDocument();
>   });
> 
>   it("renders different icons correctly", () => {
>     const { rerender } = render(<ActionMenuItem {...defaultProps} icon="Trash" />);
>     expect(screen.getByLabelText("Test Action action")).toBeInTheDocument();
> 
>     rerender(<ActionMenuItem {...defaultProps} icon="ArrowUp" />);
>     expect(screen.getByLabelText("Test Action action")).toBeInTheDocument();
>   });
> });
> ```
> 
> ✏️ Learnings added
> 🧠 Learnings used

@coderabbitai Unfortunately, with the actual code, these tests don't work. For one, <ActionMenuItem> must be a child of <ActionMenu> so the tests as they are fail. We'll need to come back to these in the future.

File: src/modules/SetListCard/components/SetSectionActionMenu/SetSectionActionMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+              icon="ArrowUp"
+              label="Move section up"
+              disabled={isInFirstSection}
+            />
+            <ActionMenuItem
+              icon="ArrowDown"
+              label="Move section down"
+              disabled={isInLastSection}
+            />
+            {setSection.position < setSectionsLength - 2 && (
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/SetListCard/components/SetSectionActionMenu/SetSectionActionMenu.tsx:113]: @coderabbitai
This is actually intentional even though `length - 2` looks weird. `length - 1` essentially means this is the last section, but the intention is that if the result of "move section to top/bottom" is the same as "move section up/down", then we'll simplify the UI by not showing that action.

The condition where this is true would be if the set section's position is second to last, which is calculated by `length - 2`.
